Plans grid: billing interval Toggle component

ciampo opened the pull request:
### Changes proposed in this Pull Request

* Add the `<PlansIntervalToggle />` component to the `@automattic/plans-grid` package, but it DOESN'T render it
* This component is heavily based on the [`<SegmentedControl />` calypso component](https://github.com/Automattic/wp-calypso/tree/trunk/client/components/segmented-control) and [how it is used in the /start flow](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-features-main/plan-type-selector.tsx)
* Compared to [how this toggle has been implemented in other flows (`/start`)](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-features-main/plan-type-selector.tsx), this implementation has a few differences:
  * the `Popover` component is imported from [`@wordpress/components`](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/popover) instead of `calypso/components/popover`. Since the new version of the component has already got its own enter transition, the `CSSTransition` component was not imported/used.
  * Since the Monthly vs Annually selection doesn't need to modify the page's URL, the `path` prop is not computed and passed to the `<SegmentedControl.Item />` component. This allowed to avoid importing the `utility-types` and `lodash` packages.
  * For translations, `@automattic/react-i18n` package was used instead of `i18n-calypso`
  * Style overrides are applied via SCSS code (instead of using `styled` components)

### Screenshots

<img width="600" src="https://user-images.githubusercontent.com/1083581/104479909-3b31c300-55c4-11eb-86e9-15d9c3a4eb0c.gif">

### Testing instructions

#### Setup

1. Open this PR's code on your local machine
2. Add the `<PlansIntervalToggle />` component to the Plans grid:
<img width="1512" alt="Screenshot 2021-01-14 at 15 25 30" src="https://user-images.githubusercontent.com/1083581/104603601-dedeaa00-567c-11eb-80d4-b1f41277ec2b.png">



#### Test in Gutenboarding

- Go through the setup steps
- Run `yarn && yarn start` in the root folder of this repo
- Visit `calyspo.localhost:3000/new` and navigate to the Plan step
- Make sure that all A/Cs from #48769 are satisfied

Fixes #48769

ciampo committed:
wip implementation of billing toggle

ciampo committed:
Copy SegmentedControl component from Calypso

ciampo committed:
WIP: rewrite Plans Interval Toggle using newly copied SegmentedControl component

ciampo committed:
Add example usage of the PlansIntervalToggle component

a8ci18n commented on packages/plans-grid/src/plans-interval-toggle/index.tsx:
```diff
@@ -0,0 +1,95 @@
+				selected={ intervalType === 'yearly' }
+				onClick={ () => onChange( 'yearly' ) }
+			>
+				<span ref={ annuallySpanRef }>{ __( 'Annually', __i18n_text_domain__ ) }</span>
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Pay annually' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/feat/plans-grid-billing-period-toggle)

ciampo commented on packages/plans-grid/src/plans-interval-toggle/index.tsx:
```diff
@@ -0,0 +1,95 @@
+ */
+import './style.scss';
+
+export type BillingIntervalType = 'yearly' | 'monthly';
```
After #48790 gets merged, this type can be removed and we can import that type from the data-store instead

ciampo committed:
Toggle: polish markup, overwrite styles

ciampo committed:
Toggle popover markup + override styles

a8ci18n commented on packages/plans-grid/src/plans-interval-toggle/index.tsx:
```diff
@@ -0,0 +1,102 @@
+					selected={ intervalType === 'yearly' }
+					onClick={ () => onChange( 'yearly' ) }
+				>
+					{ __( 'Annually', __i18n_text_domain__ ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Pay annually' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/feat/plans-grid-billing-period-toggle)

ciampo committed:
Add necessary types for SegmentedControl component

ciampo committed:
Additional instructions for translators

ciampo committed:
Remove PlansIntervalToggle test markup from plans grid

ciampo commented on packages/plans-grid/src/plans-interval-toggle/style.scss:
```diff
@@ -0,0 +1,112 @@
+@import '~@automattic/onboarding/styles/mixins';
+@import '../variables';
+
+.plans-interval-toggle {
```
In order to satisfy the design spec, I had to:
- disable the `scales/radii` rule (a `border-radius` of `5px` or `6px` is not allowed by stylelint)
- disable the `scales/font-weight` rule (`500`, aka medium font size, is not allowed by stylelint)
- hardcode the colors `#f5f5f5` and `#f2f2f2`(they are not found in [`color-studio`](https://github.com/Automattic/color-studio/blob/master/dist/color-properties.css) )

I'm open to suggestions on how to avoid all of those "exceptions"

StefanNieuwenhuis replied:
> * disable the scales/radii rule (a border-radius of 5px or 6px is not allowed by stylelint)
> * disable the scales/font-weight rule (500, aka medium font size, is not allowed by stylelint)

Why do we need to overrule them? What does it add to the UI of the toggle component? Isn't it better to keep consistent and use the default values? 

> * hardcode the colors #f5f5f5 and #f52f2f2(they are not found in color-studio )

Do we have other default values available? We might want to propose using them to stay consistent and align with default styling standards?

ciampo replied:
> Why do we need to overrule them? What does it add to the UI of the toggle component? Isn't it better to keep consistent and use the default values?

It looks like stylelint rules are not up to date with some of the new design specs

>Do we have other default values available? We might want to propose using them to stay consistent and align with default styling standards?

This is probably a question for @ollierozdarz

ciampo committed:
Remove unnecessary SCSS imports

ciampo commented on packages/plans-grid/package.json:
```diff
@@ -46,7 +46,8 @@
 		"@automattic/typography": "^1.0.0",
-		"@wordpress/base-styles": "^2.0.1"
+		"@wordpress/base-styles": "^2.0.1",
+		"prop-types": "^15.7.2"
```
Necessary because `<SegmentedControl />` is still written in JavaScript

ciampo committed:
Update BillingIntervalType to match values from plans data-store

ciampo commented on packages/plans-grid/src/plans-interval-toggle/index.tsx:
```diff
@@ -0,0 +1,103 @@
+ */
+import './style.scss';
+
+// TODO: import this type directly from plans data-store once
+// https://github.com/Automattic/wp-calypso/pull/48790 is merged
+export type BillingIntervalType = 'MONTHLY' | 'ANNUALLY';
```
Can be updated if #48790 gets merged before this PR is approved

StefanNieuwenhuis commented on packages/plans-grid/src/plans-interval-toggle/style.scss:
```diff
@@ -0,0 +1,106 @@
+
+		height: 32px;
+		min-width: 102px;
+		padding: 0 16px;
```
Why do we need padding here? And should it affect the tooltip's alignment? This is what happens when I toggle it:  

https://user-images.githubusercontent.com/12104589/104573546-4da80d00-5655-11eb-8cca-ced66acea501.mov

ciampo replied:
Edit: I slightly misread your initial comment here

The padding is mostly there because the toggle items have a minimum width of `102px`, but if the text gets longer (for example, when translated to another language), that padding is there to ensure some white space around it

As you pointed out, is also affects the position of the popover. It's hard to replicate exactly the same positioning and spacing of the `/start` flow, because I had to use a different Popover component (as explained in the PR description).

<img width="606" alt="Screenshot 2021-01-14 at 10 48 55" src="https://user-images.githubusercontent.com/1083581/104574257-1b4adf80-5656-11eb-8a53-fd8fa4b7f4d5.png">


I'll see if I can move the padding to another element, which may help pushing the popover to the right

StefanNieuwenhuis replied:
Thanks for elaborating! It might make sense to move the padding to its parent `<a/>` since this is already a block element which might makes life (regarding padding) a bit easier?

StefanNieuwenhuis replied:
Sorry for the spam but why do we need a `min-width` and fixed `height`? And how does that affect the (mobile) viewports?

ciampo replied:
The `min-width` and `height` are following the design specs.

In particular for the min-width, both toggle "items" in the design have the same width, even if they don't have the same text. So I decided to add a `min-width` rule, which will make the items the same width, unless the copy gets longer

ciampo replied:
I tried moving padding to the `<a>` parent, but the positioning of the popover doesn't change. This is because the Popover gets anchored to the `<span>` element, and so, moving the padding from the `span` to the parent `a` doesn't affect the actual width of the `span` contents.

The only way to push the popover position more to the right is to remove the `padding` completely, but that will have negative effects on how longer labels will look like in the toggle (in this example, notice the lack of space to the left of "Some" and to the right of "here")

<img width="568" alt="Screenshot 2021-01-14 at 12 17 08" src="https://user-images.githubusercontent.com/1083581/104584195-7be01980-5662-11eb-8815-efddb92e6b83.png">

ciampo replied:
Update: I actually managed to have this working by wrapping the content of the Segmented Control Items into an additional span. This is probably achieving the best of both worlds

<img width="891" alt="Screenshot 2021-01-14 at 15 36 35" src="https://user-images.githubusercontent.com/1083581/104605150-6b3d9c80-567e-11eb-82ba-8e4f722b9d98.png">

StefanNieuwenhuis commented:
The `<PlansIntervalToggle/>` has no default selected option. Is this intentional? 
<img width="251" alt="Screenshot 2021-01-14 at 11 07 00" src="https://user-images.githubusercontent.com/12104589/104576802-f4da7380-5658-11eb-891e-4a07e04efd57.png">

StefanNieuwenhuis commented on packages/plans-grid/src/plans-grid/style.scss:
```diff
@@ -20,6 +20,14 @@
 }
 
+.plans-grid__toggle {
+	margin: -24px 0 32px;
+
+	@include break-medium {
+		margin-bottom: 40px;
+	}
+}
```
Are we planning to reuse the `PlansIntervalToggle`? If that's the case, it's better to add the margins to the `.plans-grid__table`. That prevents the component from being too opinionated.

ciampo replied:
There's a few reasons I went for this approach — one is, in case we ever remove the toggle, you'd only need to delete the toggle styles.

Also, the toggle has a negative margin top — which means that the alternative would be to change the margin bottom of `.plans-grid__header`. But [that margin comes from the `onboarding-heading-padding` mixin](https://github.com/Automattic/wp-calypso/blob/ce46322677bfbcdea6e443392982a2fa14c2f9d3/packages/plans-grid/src/plans-grid/style.scss#L17), which is reused across a lot of other components. So I decided to leave that untouched, and use a negative margin on the toggle instead.

StefanNieuwenhuis replied:
It makes total sense to reuse as much as possible so it makes sense to do that here as well. Thanks for explaining 👍

ciampo commented:
> The `<PlansIntervalToggle/>` has no default selected option. Is this intentional?
> <img alt="Screenshot 2021-01-14 at 11 07 00" width="251" src="https://user-images.githubusercontent.com/12104589/104576802-f4da7380-5658-11eb-891e-4a07e04efd57.png">

That’s not supposed to happen — the <PlansIntervalToggle/> component requires the `intervalType` prop to be defined, which should set the initial selection too. Can you double-check if you're passing that property correctly, when you manually added the component to your local version for testing it? You should use [this commit](https://github.com/Automattic/wp-calypso/pull/48800/commits/76fab1c13282921d4ab735a1f7ee9c48dd544c97) as a general guide, although you may need to change the initial value of `billingInterval` from `'yearly'` to `'ANNUALLY'`, following the latest changes in [this commit](https://github.com/Automattic/wp-calypso/pull/48800/commits/ce46322677bfbcdea6e443392982a2fa14c2f9d3)

ciampo committed:
Fix popover positioning by wrapping segmented control items in one additional span

a8ci18n commented on packages/plans-grid/src/plans-interval-toggle/index.tsx:
```diff
@@ -0,0 +1,107 @@
+					onClick={ () => onChange( 'ANNUALLY' ) }
+				>
+					<span className="plans-interval-toggle__label">
+						{ __( 'Annually', __i18n_text_domain__ ) }
```
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **20** times:
`translate( 'Pay annually' )` *ES Score: 8*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/feat/plans-grid-billing-period-toggle)

StefanNieuwenhuis commented:
> The `<PlansIntervalToggle/>` has no default selected option. Is this intentional?
> <img alt="Screenshot 2021-01-14 at 11 07 00" width="251" src="https://user-images.githubusercontent.com/12104589/104576802-f4da7380-5658-11eb-891e-4a07e04efd57.png">

FIXED! The `billingInterval` value (copied from the other PR in the testing instructions) was wrong. Changing to the correct value fixed it.

ciampo commented:
Rebased on top of trunk, which requires node `12.20.1`

Commit:
Add import to prevent launch flow from breaking inside ET.

razvanpapadopol commented:
Tested inside ET as well. Focused Launch was crashing before https://github.com/Automattic/wp-calypso/pull/48800/commits/e9066faf3bb8f9af86975eda32107ec7d1720e97 but even after the fix, UI is still looking a bit off:
<img width="390" alt="Screenshot 2021-01-14 at 18 15 28" src="https://user-images.githubusercontent.com/14192054/104617986-d003f180-5694-11eb-9694-8977bec6e9df.png">

ciampo committed:
Override more SegmentedControl styles

This it to protect from additional CSS rules that come into play
when this component is used in the Block Editor (via gutenframe — e.g
in the Focused Launch)

ciampo committed:
Override more Popover styles

This it to protect from additional CSS rules that come into play
when this component is used in the Block Editor (via gutenframe — e.g
in the Focused Launch)

ciampo commented:
> Tested inside ET as well. Focused Launch was crashing before [e9066fa](https://github.com/Automattic/wp-calypso/commit/e9066faf3bb8f9af86975eda32107ec7d1720e97) but even after the fix, UI is still looking a bit off:
> <img alt="Screenshot 2021-01-14 at 18 15 28" width="390" src="https://user-images.githubusercontent.com/14192054/104617986-d003f180-5694-11eb-9694-8977bec6e9df.png">

Hey @razvanpapadopol, the component should now look fine also in the Focused Launch flow

razvanpapadopol reviewed (approved):
I can confirm the toggle is visually consistent across browsers (Chrome, Firefox, Safari) and environments (Gutenboardign and inside iframe in Launch flow) 👍

razvanpapadopol commented on packages/plans-grid/src/plans-interval-toggle/index.tsx:
```diff
@@ -0,0 +1,107 @@
+	);
+};
+
+type ToggleHostProps = {
```
nit: we usually use `interface` for props typing

ciampo replied:
I'll change this in the next PR

razvanpapadopol commented on packages/plans-grid/src/plans-interval-toggle/index.tsx:
```diff
@@ -0,0 +1,107 @@
+type ToggleHostProps = {
+	intervalType: BillingIntervalType;
+	onChange: ( selectedValue: BillingIntervalType ) => void;
+	maxSavingsPerc: number;
```
I believe we are using `discount` in https://github.com/Automattic/wp-calypso/pull/48790 and `Perc` doesn't feel to readable. 
So what about changing this to `maxDiscountPercentage` ?

ciampo replied:
Will also be changed in the next PR
